[CGP] Undo constant propagation of pointers across calls

antoniofrighetto opened the pull request:
It may be profitable to undo SCCP propagation of C++ static values, if such constants are pointers, in order to avoid redundant pointer computation, since the method returning the constant is non-removable.

antoniofrighetto commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2555,6 +2555,37 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+    });
+  };
+
+  auto *PrevCI = dyn_cast_or_null<CallInst>(CI->getPrevNonDebugInstruction());
```
Updated, thanks for the suggestion!

antoniofrighetto commented:
Uhmm, it seems like there's a failure when building llvm-min-tblgen in stage2 (from the compile time tracker), I'm taking a look.

dtcxzyw commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2555,6 +2555,45 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+  if (Constant *RV = GetUniformReturnValue(CI->getCalledFunction())) {
+    bool MadeChange = false;
+    const auto &DT = getDT(*CI->getFunction());
+    for (Use &U : RV->uses()) {
```
```suggestion
    for (Use &U : make_early_inc_range(RV->uses())) {
```

antoniofrighetto replied:
Right, updated, thanks!

dtcxzyw commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2555,6 +2555,45 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+  // computation of the constant, as the function method returning the constant
+  // needs to be executed anyways.
+  auto GetUniformReturnValue = [](const Function *F) -> Constant * {
+    Constant *UniformValue = nullptr;
+    for (auto &BB : llvm::reverse(*F)) {
+      if (auto *Ret = dyn_cast<ReturnInst>(BB.getTerminator())) {
+        if (auto *Val = dyn_cast_or_null<GlobalVariable>(Ret->getReturnValue());
```
```suggestion
   if (!F->getReturnType()->isPointerTy()) return nullptr;

    Constant *UniformValue = nullptr;
    for (auto &BB : llvm::reverse(*F)) {
      if (auto *Ret = dyn_cast<ReturnInst>(BB.getTerminator())) {
        if (auto *Val = dyn_cast<GlobalVariable>(Ret->getReturnValue()) {
```

It should be clearer :)

goldsteinn commented on llvm/test/Transforms/CodeGenPrepare/remove-constant-on-calls.ll:
```diff
@@ -0,0 +1,54 @@
+declare i32 @S_getI(ptr nonnull align 1 dereferenceable(1))
+declare void @__cxa_guard_abort(ptr)
+declare void @__cxa_guard_release(ptr)
+declare i32 @__gxx_personality_v0(...)
```
Can you pre-commit the test?

dtcxzyw commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2686,47 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+    Constant *UniformValue = nullptr;
+    for (auto &BB : llvm::reverse(*F)) {
+      if (auto *RI = dyn_cast<ReturnInst>(BB.getTerminator())) {
+        if (auto *V = dyn_cast_or_null<GlobalVariable>(RI->getReturnValue())) {
```
```suggestion
        if (auto *V = dyn_cast<GlobalVariable>(RI->getReturnValue())) {
```

dtcxzyw reviewed (approved):
LGTM.

nikic commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2686,47 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
     return true;
   }
 
+  // SCCP may have propagated C++ static variables across calls. If this happens
```
This doesn't really seem specific to C++ static variables.

nikic commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2686,47 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+      return nullptr;
+
+    Constant *UniformValue = nullptr;
+    for (auto &BB : llvm::reverse(*F)) {
```
Why the reverse iteration here?

nikic commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2686,47 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+    return UniformValue;
+  };
+
+  if (Constant *RV = GetUniformReturnValue(CI->getCalledFunction())) {
```
This is incorrect for interposable functions.

antoniofrighetto replied:
Right, thanks!

nikic commented on llvm/test/Transforms/CodeGenPrepare/revert-constant-ptr-propagation-on-calls.ll:
There is a lot of test coverage missing here (all the negative tests).

antoniofrighetto replied:
Negative tests added.

nikic commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2686,47 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+    const auto &DT = getDT(*CI->getFunction());
+    for (Use &U : make_early_inc_range(RV->uses())) {
+      auto *I = dyn_cast<Instruction>(U.getUser());
+      if (!I || I->getParent() != CI->getParent())
```
Is `I->getParent() != CI->getParent()` a profitability heuristic, to avoid long live ranges? If so, add a comment please.

antoniofrighetto committed:
[CGP] Introduce test for PR102926 (NFC)

antoniofrighetto committed:
[CGP] Undo constant propagation of pointers across calls

It may be profitable to revert SCCP propagation of C++ static values,
if such constants are pointers, in order to avoid redundant pointer
computation, since the method returning the constant is non-removable.

nikic commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2687,52 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+    return UniformValue;
+  };
+
+  if (!Callee->isInterposable() && !Callee->hasDLLImportStorageClass()) {
```
Why is the hasDLLImportStorageClass() check needed?

antoniofrighetto replied:
Uff, mixed up available_externally with extern_weak.

nikic commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2687,52 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+  if (!Callee->isInterposable() && !Callee->hasDLLImportStorageClass()) {
+    if (Constant *RV = GetUniformReturnValue(Callee)) {
+      bool MadeChange = false;
+      const auto &DT = getDT(*CI->getFunction());
```
If we're in the same BB, you may as well skip the dominator tree and use comesBefore() instead.

nikic commented on llvm/test/Transforms/CodeGenPrepare/revert-constant-ptr-propagation-on-calls.ll:
```diff
@@ -0,0 +1,123 @@
+; CHECK-NEXT:    [[GETI:%.*]] = call i32 @S_getI(ptr nonnull align 1 dereferenceable(1) [[GETS_PTR]])
+; CHECK-NEXT:    ret i32 [[GETI]]
+;
+  %getS_ptr = call nonnull align 1 dereferenceable(1) ptr @getS()
```
Remove unnecessary attributes here and elsewhere.

nikic commented on llvm/test/Transforms/CodeGenPrepare/revert-constant-ptr-propagation-on-calls.ll:
```diff
@@ -0,0 +1,123 @@
+  %getS_ptr = call nonnull align 1 dereferenceable(1) ptr @getS_dllimport_function()
+  %getI = call i32 @S_getI(ptr nonnull align 1 dereferenceable(1) @g_getS)
+  ret i32 %getI
+}
```
Add test with use in other block and use before call.

nikic commented on llvm/test/Transforms/CodeGenPrepare/revert-constant-ptr-propagation-on-calls.ll:
```diff
@@ -0,0 +1,123 @@
+
+declare ptr @getS_dec()
+
+define nonnull align 1 dereferenceable(1) ptr @getS() personality ptr @__gxx_personality_v0 {
```
We should also have coverage for the logic that detects the return value, e.g. a function with two different return values as a negative test.

antoniofrighetto replied:
Should be taken into account in test function `caller_4`.

nikic commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2687,51 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+    return UniformValue;
+  };
+
+  if (!Callee->isInterposable()) {
```
```suggestion
  if (!Callee->hasExactDefinition()) {
```
I think we should check for an exact definition here, to avoid any derefinement issues. SCCP will not propagate the return value without exact definition anyway.

(Consider a function that returns a global or poison and the visible definition folds to the global, while an ODR equivalent returns poison on one path.)

nikic replied:
You can test this using a linkonce_odr function.

antoniofrighetto replied:
Agree, thanks, was indeed previously dubious on this as available_externally was not contemplated by isInterposable.

nikic commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2686,6 +2687,51 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
+  };
+
+  if (!Callee->isInterposable()) {
+    if (Constant *RV = GetUniformReturnValue(Callee)) {
```
It would be better to make the return value GlobalVariable. Iterating the use list of arbitrary Constants is iffy.

nikic reviewed (approved):
LGTM
